crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -382,10 +382,16 @@ impl Acir {
 
     //Map the outputs into the array
     fn map_array(&mut self, a: ArrayId, outputs: &[Witness], ctx: &SsaContext) {
-        let adr = ctx.mem[a].adr;
-        for i in outputs.iter().enumerate() {
-            let var = InternalVar::from(*i.1);
-            self.memory_map.insert(adr + i.0 as u32, var);
+        let array = &ctx.mem[a];
+        let adr = array.adr;
+        for i in 0..array.len {
+            if i < outputs.len() as u32 {
+                let var = InternalVar::from(outputs[i as usize]);
+                self.memory_map.insert(adr + i, var);
+            } else {
+                let var = InternalVar::from(Expression::zero());
+                self.memory_map.insert(adr + i, var);
```

Comment:
Would there be a difference in behavior if we only looped over `0 .. outputs.len()` instead? E.g. will these extra `0` elements actually be accessed later on and could that affect the output as well?

Reply:
There would be a difference as this was essentially what was happening before. This really only affects mapping of the outputs for `to_bits` and `to_bytes` where the outputs may be smaller than the array being outputted from the functions. In the case of `to_bytes` if we specify a smaller byte size like 31 (which we force the user to do for safety atm), `outputs.len()` will be `31` and we will only map `0 .. 31`. For a byte size of `30`, `outputs.len()` will give a range of `0 .. 30`, etc. These `0` elements will be accessed later on and this actually why I added this if statement. If a user tries to access elements returned from `to_bits` or `to_bytes` past the bit/byte size specified when calling the function they will get an unclear panic in the acir gen. The most prominent annoyance I see with this is when a user tries to pass an entire bit/byte array to another function like [here](https://github.com/vezenovm/aztec-circuits-noir/blob/d952abb4f1110c391fda01ec8019056d2ad6d6fe/circuits/join_split/src/main.nr#L200). Technically users could avoid this panic through hack-y methods, but I felt we should add this to make sure we are mapping the entire array correctly, including for the trialing zero values of the array.

Reply:
I could add a comment to elaborate if that would make this choice more clear

Reply:
I think it makes sense - I had missed that we were essentially already looping to outputs.len() in the version with the bug.